### ISSUE TITLE
Discussion about SPRT bounds.

### ISSUE THREAD

**Issue by Vizvezdenec** [glinscott/fishtest#764]
After first days of huge elo gains with NNUE everything more or less calmed down so I think it's about time to start discussion of possible change in SPRT bounds.
With NNUE integration our drawrate at fishtest really skyrocketed (no contempt for NNUE also doesn't help ofc). So it makes even harder to catch any improvements, especially at LTC, since for this strong of a play even 1 elo is kinda huge.
On the other hand with this draw rate tests started to converge much faster - prior to NNUE we had 75%~ drawrate at LTC, currently it's 90% which makes LTC converge 2.5 times faster.
So it will be logical to lower SPRT for LTC even further to catch more improvements, smth like [0,25 ; 1,25] (or 0,2 ; 1,2 ) will make LTC SPRT converge in the same ~ number of games as previously but will allow to catch more patches and (as a side result) will also increase confidence in passing test being non-negative.
I'm not saying this should be done now but, imho, this is eventually what we will need to do, current hitrate of green STC getting a green LTC is kinda abysmal in latest days :) 
Well, it's not an immediate measure, but rather a discussion of smth that eventually should probably be done.
What do you think? @vondele 
All calculations were made using fishtest-integrated SPRT calculator.

**Comment by vondele** [comment]
yes, I've been thinking about this. Need to experiment a bit. Right now a patch like https://tests.stockfishchess.org/tests/view/5f3a58f6b38d442594aabe0c has a W/L ratio that is roughly 1.05, which seems like worth having, but it is at 1Elo, and might thus be rejected. So, reducing a little the upper bound might make sense. We might also have to update the simplification bounds, certainly if we adjust the passing Elo threshold to 0.75. 

Maybe @vdbergh has an opinion.

**Comment by vondele** [comment]
Some, some experimenting from my side. First draw rates:

|         | classical | NNUE |
| ---    | --- | --- |
| STC | 61% | 78% |
| LTC  | 75% | 89% |

I wonder if a new net could gain a few Elo, but completely change these stats. A 'contemptuous net' for example. A single merge could, in that case, require a re-adjustment of bounds discussed below.

Based on these draw rates, the current bounds have roughly 100k an expected maximum number of games, while with NNUE it is roughly half of that. If we want to get back to ~100k games max, while making some adjustments so that LTC tests pass earlier, 0 Elo patches have a harder time, and simplifications remain as easy for 0Elo, but harder for -0.5 Elo, we could use
Gainers: STC {-0.25, 1.25}, LTC {0.25, 1.25}  Simplifications: STC {-1.25, 0.25} LTC {-0.75, 0.25}. 

@Vizvezdenec what do you think?

**Comment by Vizvezdenec** [comment]
Well, I played with SPRT calculator a bit.
Previously at LTC we had 116,3k games average converging at max and 29k at leafs, now 51,2k at max and 13k at leafs, so average number of games dropped by 2.28.
Probability of patch passing LTC and having <= 0 elo is 1,9% (unchanged).
We want to kinda remain this stats (I guess?).
For this purposes I think LTC should be {0.2, 1.2} - we will have this peak at 115,1k (so almost the same as previously), probability of negative test passing 1,6% - which is even better.
I don't really think that STC needs this much of a change honestly since we don't have real trouble in getting passing STCs :) Maybe it can be reconsidered later.
About simplifications it's hard to tell, I was never really involved in a lot of simplifying. They for sure need to be slightly stricter, although I don't know about them being assymetrical tbh. {-1.0, 0.5} for both? This is an object for discussion for sure, maybe we should prohibit simplifications for now a bit.
So TLDR - I think following
Gainers {-0.5, 1.5} / {0.2, 1.2}, Simplifications - both {-1, 0.5}.
Maybe we should summon more people to discuss this from active developers list?
@Lolligerhans @unaiic @SFisGOD @locutus2 @MJZ1977 @mstembera @pb00068

**Comment by unaiic** [comment]
I agree to lower the upper bound in LTC and have stricter simplification bounds, specially now that NNUE could have changed the effectiveness of some terms. As @Vizvezdenec pointed out, I don't think STC bounds need much change (we continue passing them quite easily). Any of the bounds for both STC and LTC (ELO gainers) you have proposed seems fine to me, keeping the LTC bounds at least {0.2, 1.2} or higher {0.25, 1.25}. Simplification bounds, however, should be stricter than {-1.0, 0.5} IMHO; I'd prefer @vondele's STC {-1.25, 0.25} LTC {-0.75, 0.25} bounds, even though I'd prefer them being symmetrical.

**Comment by vondele** [comment]
note that STC {-0.25, 1.25} isn't easier to pass than STC {-0.5, 1.5} (at least same 50% chance at 0.5 Elo), just less lucky passers.

**Comment by Vizvezdenec** [comment]
With 0.2; 1.2 we will get less lucky passers for LTC by like 19% (1,9% -> 1,6%), do we even need to have less lucky passers for STC also since we also know that a lot of search stuff is not really linear (or at least some of it?).
Shrinking STC bounds would've made more sense if we had really low numbers of passing STCs but so far it's not really the case. At least for now I would consider not changing STC bounds at all.

**Comment by vondele** [comment]
lucky passers cost a lot of resources (roughly 10x more for an LTC with proposed bounds vs STC with current bounds), so we should try to minimize those. The passing probability is also symmetric around the mid-point, so a wide interval to have lucky passers, means also that we have unlucky tests that fail. To adjust what we let through it is the midpoint of the interval that has to shift, while it the interval itself should be made as narrow as possible while keeping the resource usage reasonable.

**Comment by Vizvezdenec** [comment]
Actually...
Do we still recalibrate TC to match 60+0.6 with 1,5 mnps?
Because average fishtest speed dropped by like 30-40%. If we do so then machines that had 1,5 mnps are currently running 90+0.9 or smth similar instead of 60+0.6 they did run previously, maybe this is also causing a huge increase in draw rate?
If so than maybe we should lower actual time controls used?

**Comment by Vizvezdenec** [comment]
I just checked - my laptop is running fishtest at 22+2.2 for STC (!!!).
No wonder draw rate is so high - we effectively have x1.5 average time control increase for both STC and LTC!

**Comment by vondele** [comment]
yes, I'm aware of this, we scale TC with an expected 1.6Mn which some workers don't get these days (notice for the mixed bench). I was thinking this wasn't too bad, since increasing TC is a common request as well.

**Comment by Vizvezdenec** [comment]
Well idk I think it's too high now.
Effectively we had like 1,4 mnps on average / core, now it's <= 1 mnps. I don't think it's all that good, especially since draw rate is getting so high.
Increasing TC was a common request but it wasn't really fulfilled for a reason :) Currently we are seeing this reasons - simplifications are really easy to pass (and this is what probably causes RT regression now), tests are REALLY hard to pass, especially LTC, because of drawrate really increases a lot.
Also I don't remember for anyone asking to increase STC also :)
I think falling back to smth like 7,5 +0.075 & 45+0,45 will be a good thing - maybe we wouldn't even need SPRT changes at all and this will still be >= of what we had in terms of time control (for 80%+ of fishtest at least). Or recalculate to "normal" speed 1,2 mnps instead of 1,6 (which is more or less the same).

**Comment by vondele** [comment]
as an additional consideration, most of the workers are hyperthreaded (noob), which probably has a different effect with classical and NNUE, which affects the effective TC as well. Probably that is accounted for in the scaling currently.

**Comment by Vizvezdenec** [comment]
Well we didn't have scaling issues with first NNUE merge...
I think it's smth to inverstigate, but it's not really related to current issue, imho :) 
Also I have to note that prior to NNUE integration we already had 0,5 elo / elo gainer on average, now with increased TC and no contempt that shrinks elo gains and overall stronger engines this 0,5 elo may be even lower and almost impossible to detect.  
My opinion is that steps should be the following:
1) slight lower of TC in form of either TC lowering or recalculating to lower normal speed;
2) only then collecting data about draw rates and maybe adjustment of SPRT to match what we previously had in terms of negative passing probability and average number of games / patch.

**Comment by unaiic** [comment]
I agree with @Vizvezdenec. I think those are sensible steps to take. We could start with the TC and only then see if adjusting SPRT is necessary.

**Comment by vondele** [comment]
probably we need to adjust the reference nps, but it is similar. We'd like to have TC roughly (on average) reflect the real TC. This probably needs to be done after a release, e.g. RT will change quite a bit if we change TC.

**Comment by Vizvezdenec** [comment]
Will it though? I don't think that RT will change THAT much from x1.5 time / game.
What sure changes is dynamics of SPRT cause it depends on draw rate a lot, but I remember that we did some RTs with 180+1.8 and they were within error bars from 60+0.6 every time, like 2 elo from 30 less ~~

**Comment by vondele** [comment]
So, before we change the default, could you just start a few tests with 1.5x less time, and we see what it does to the draw rate?

**Comment by vondele** [comment]
I guess the draw rate comes out pretty quickly, e.g. 10000 games, and we can test a few TCs to have a feeling for how the draw rate changes. Maybe it is better not to run identical binaries, but e.g. the the versions corresponding to the two last RT? That would also give an idea about the actual Elo difference between these two versions, and complement the number we're getting right now.

**Comment by Vizvezdenec** [comment]
I will have easier times to do this with master branch than with some older binary I will need to create in my repo honestly :D 
Also I think if we are measuring it for positive SPRT, we are measuring it for "almost neutral" patches usually (like you can let's say delete eval and have 0% draw rate, but it mean nothing), so master vs master will do.
I will run 10k games in both 7,5 + 0,075 and 45 + 0,45 master vs master to see what draw rates we will get.

**Comment by vondele** [comment]
OK, fine.

Note that you don't need to create branches for RT1 vs RT2 testing. In the branch field you just put the SHA of the commit you want to test, and it will be found in your repo (because your master branch contains those all).

**Comment by vondele** [comment]
FYI @lp--

**Comment by MJZ1977** [comment]
We can assume that with NNUE, SF is stronger at TC/2 than with classical eval at TC. 
So is seems reasonable to make test at half TC with more draw rates and more games per test.

**Comment by Vizvezdenec** [comment]
Okay, tests finished, so some data.
60+0.6 drawrate 90,7%
45+0.45 draw rate 88,1%
10+0.1 draw rate 78,1%
7,5 + ... draw rate 76,4%
the huge spike of number of decisive games (negative one) is in 60+0.6, 11,9% for 45 seconds and 9,3% for 60 seconds, it's 28%
Difference between 10 and 7,5 is not that big.
So if anything it affects LTC more thus giving us even more trouble catching anything here.
I think that we should recalculate to 1,2 mnps, stcs will be like 20% faster with it and LTCs will be better at catching elo gainers, because currently they are almost impossible to pass and it's not looking any better with engine getting stronger and stronger.
About SPRT bounds if we switch to 1,2 instead of 1,6 baseline.
for LTCs we had 75% draw rate, 1,9% of negative pass and 116,3k games needed at "worst" possible elo to converge.
with 88,1% draw rate to maintain the same stats we need smth like {0.175, 1.2} (yes, I know it looks ugly :D ) which will have 88,1% draw rate, 1,9% of negative pass and 118,5k games at peak. It will lower 50% elo from 1 elo to 0,7 elo. Maybe to it as {0.2, 1.2} to slightly decrease probability of negative pass, number of needed games will increase to 124,5k. Other variant can be {0.2, 1.25} - this will be 1,7% negative pass and 112,8k games at peak, but 50% elo will increase from slightly less than 0,7 elo to 0,72 or smth similar.
For STCs we had 102k peak games, negative probability of 18,7%.
I think that decreasing STC to 7,5 (or rather recalculating to lower baseline speed) wouldn't hurt anything - STC is pretty easy to pass and also draw rate wouldn't be too low as my test shows, so tests wouldn't take ages to complete.
So with new 76,4% draw rate to maintain ~ the same stats of 102k peak games and negative probability of 18,7% bounds should be smth like {-0.4, 1,2} - this will maintain the same negative % but lower number of peak games to 96,5k, also we will have 50% elo decreased from 0,5 elo to 0,4 elo.
Other variant can be {-0.35, 1.2}.
So overall if we change 
baseline speed 1,2 mnps
STC SPRT {-0.4, 1.2}
LTC SPRT {0,2, 1.2}
We will achieve: 
Effective TC approximately back to what it was before NNUE.
STC will have number of games needed to pass decreased by 5,5%, probability of negative pass will be the same, elo of patch to pass STC with 50% probability is 0,4 elo.
LTC will have number of games needed to pass increased by 7%, probability of negative pass lowered from 1,9% to 1,6%, elo of patch to pass STC with 50% probability is 0,7 elo.
So with this average patch that wouldn't go to LTC will converge 5,5% faster, patch that will go to LTC will be like 5% slower (number is hard to calculate tbh), elo of patch passing BOTH SPRTs with 50% probability will decrease from 1,1 elo to 0,8 elo so by 30% (actually slightly more, it's rounding).
About simplifications I can do ~ similar recalculations but they will be simplier because their bars are symmetrical. BUT I think that it should be done differently since if we decrease 50% passrate elo by 30% we should do the same for simplifications - so make 50% passrate of a simplification be decreased by 30% prohibiting simplifications from eating all elo gains.

**Comment by Vizvezdenec** [comment]
Note that changing baseline to 1,2 mnps will makes EVERY test converge 25% faster because of lowering of their effective time controls.

**Comment by MJZ1977** [comment]
TC 7.5s and 45s are little strange. Why not simply 5s and 30s. It will not make a big difference I think ...

**Comment by Vizvezdenec** [comment]
Well the main reason is that on average fishtest nps is dropped by like 25-40% since NNUE is innately slower. So CPUs that were running at 60+0.6 now run on smth like 80+0.8 or 90+0.9.
45+0.45 is a way for them to run at 60+0.6 at they did previously.
But I propose not to change TC but to change baseline speed, nowadays it's recalibrated to 1.6 mnps - I propose 1.2 mnps to keep baseline / fishtest average like 1,1-1,2 as it was previously.
This wouldn't change TC that we see on fishtest but will change EFFECTIVE TC which is used in actual testing games by 25%. 
To calculate draw rate of this I used 45+0.45 and 7.5+0.075 because it's basically the same.

**Comment by Vizvezdenec** [comment]
So, about simplifications.
We had 61% drawrate STC, 75% drawrate LTC, 
50% of pass for each of them was -0.5 elo and number of games was 102k/65,4k for STC/LTC.
With new proposed bounds 50% pass elo decreased by 30%, so it's logical to do the same for simplifications, so make it 0,35 elo.
Number of games should be approximately the same.
So proposed bounds are {-1.1, 0.4}, their stats are the following:
50% pass elo is -0.35 elo, for STC (with proposed baseline change) number of games is 109.8k, for LTC - 55,4k.
So STC will converge like 7% slower but LTC approximately 10% weaker, 0 elo patch probability slightly lowered, 50% elo pass is matching proposed gaining SPRTs -
previously we had 50% pass probability for BOTH SPRTs being approximately -0.2 elo, now it will be smth like -0.125 elo, so lowered by ~ the same % in order for simplifications to not eat all gains + tests should converge slightly faster because of faster LTCs.

**Comment by Vizvezdenec** [comment]
so TLDR for all this, proposed changes:
baseline nps - 1.2 mnps
elo gainers - 
STC SPRT {-0.4, 1.2}
LTC SPRT {0,2, 1.2}
simplifications -
{-1.1, 0.4} both

**Comment by vondele** [comment]
I agree with nps -> 1.2 .. however, that's best reset after a SF12 release, just in order not to make the RT more confusing.

The bounds, I agree with making Elo gainers somewhat easier. I still like STC {-0.25, 1.25}, LTC {0.25, 1.25} but it is similar to what you propose. Except for pushing a little more on STC confidence. Maybe that's not needed.

The simplifications need to be stricter IMO. I had STC {-1.25, 0.25} LTC {-0.75, 0.25}. What seems to happen now two simplifications that can seemingly be introduced without changes back is not good. I think we need to have the LTC interval narrow enough so we have higher confidence in the result.

**Comment by Vizvezdenec** [comment]
We can just make RTs 80+0.8 and 40+0.4 with new baseline and be good with it - it wouldn't be a big problem since there are like 2 people who submit them. Just keep them informed and be gucci :)
I recalculated simplifications in order for them to match increased elo catching ability of elo gainers. So patch that had previously 50% to pass will have 30% or so with new bounds, and overall needed elo for 50% pass rate is lowered by the same 30% that elo gainers elo lowered.
We can maybe do some experiments - make some simplifications (some people will do it :P ) and then test for reintroduction, it's let's say 5 test fail we can conclude that new bounds are good enough, if not make them even more strict.
I don't really like such a strict LTC for simplifications because it will really skyrocket needed games to pass LTC - my proposal has peak of 55,4 k games, current bounds with previous draw rate - 65,5k games, your will be 124.5k games - it will cause simplifications to eat a ton of resources which is probably not a good thing to do. After all, we are there to improve SF for the most time :)

**Comment by Vizvezdenec** [comment]
Also side note - I don't like bounds being proposed "by feel". We had this with previous bounds (prior to current) and it kinda led to disaster. I prefer to calculate what they bring and what they take.

**Comment by unaiic** [comment]
I agree making simplifications slightly stricter. We should be confident that a code removal/simplification won't affect lowering ELO. Regarding standard STC and LTC, however, I have no complaints; both options look good to me. But it is true that tests will tell us more than mere feelings or good-looking numbers.
I currently have some simplification tests running with proposed LTC {-0.75, 0.25} bounds. Maybe they could help...

**Comment by Vizvezdenec** [comment]
The problem with this bounds is that they will eat much more resources than what we currently have. I don't think that we want to spend this much on simplifications, currently to converge at LTC they take half of the games on "normal" patch, while with this strict bounds they will take the ~ same number of games - I really don't like it because our priority should be on elo gainers mostly, simplifying is nice but not as a main course.

**Comment by vondele** [comment]
I of course use the SPRT calculator to estimate. I'm trying to have the same confidence in simplifications as we have in Elo gainers.

**Comment by Vizvezdenec** [comment]
well what do you define as a "confidence"? :) 
I used idea that 50% pass elo should drop by the same value (approximately) that we had it dropping for elo gainers. Well maybe we can make them slightly stricter than what I proposed, but I think having {-0.75, 0.25} is a bit of an overkill in terms of performance. Imho we don't want this long LTCs for simplifications.
Maybe go for some intermediate value like {-0.9, 0.4}? (and whatever at STC) This should be good balance between confidence/not occupying fishtest.

**Comment by vondele** [comment]
'confidence' (as I mis-use) is basically related to the width of the interval.  The midpoint is the threshold of what we would like to accept and what we would like to reject (this value always has 50% prob). The width of the interval defines 'the grey zone'. I like to have the grey zone similar for both Elo gainers and simplifications, so we know they are close to the threshold in both cases. Furthermore, it is nice if the intervals are not overlapping, since if they overlap, patches in that area can with ~5% chance pass both tests. For the same width of interval all tests at the same TC roughly take the same number of games.

**Comment by Vizvezdenec** [comment]
I have an issue with having the same width of the interval for at least LTC of simplifications and elo gainers because it means that simplifications will take the same number of games as elo gainers to converge.
I think we can use slightly more loose bounds (with the same midpoint ofc) just for them to use less resources as slightly less important tests overall. 
But I'm overall okay with stricter bounds. I think we should force elo gainers bounds change as a first priority. 
I really like changing baseline because it has quite huge impact on LTC, sure, it will slightly mess up RTs which we can "fix" by just adjusting TC for them, we should inform alayant if we do so and it will probably be pretty much okay.
As for SPRT bounds it's up for you to choose, I proposed my numbers, I think (honestly) they are slightly better than yours, they increase probability of test being positive compared to what we have now (and now I think it works fine) + allow to catch much more elo gainers while returning to average converging times we had before NNUE.
So overall what will be the conclusion? I think we discussed enough by this point :)

**Comment by vondele** [comment]
Baseline change requires a worker update, including noob's fleet to be done manually. So that's the least easy one to make. It makes 25% difference. Maybe we can do it quite soon, I don't think we should delay a release too much by now (but needs talking to other people). While you might call simplifications less important, they are typically more controversial, to have fewer discussions (and waste human time), I like them to be passing with confidence. Changing the bounds is rather easy, a patch to fishtest, and server update, so that is likely going to happen first. Since I would like the branch to become stable for a release, I actually wouldn't mind if the patch rate goes down temporarily.

**Comment by Vizvezdenec** [comment]
Understandable. 
I think we should make PR about bounds then :)

**Comment by lp--** [comment]
So what are proposed probabilities for simplifications for -1, -0.5, 0 elo patches for STC, LTC and expected number of games?

**Comment by lp--** [comment]
[Here there was my proposition](https://github.com/official-stockfish/Stockfish/pull/3024#issuecomment-676785532) for simplificaton  bounds

Improvement bounds seems ok for now. But if rate of successful patches starts to drop I was considering [0.2, 1.4] bounds which with new draw rate has approximately the same characteristics as they were before with old one.

**Comment by Vizvezdenec** [comment]
so @vondele do you want me to create a PR?
If so, what bounds for STC/LTC/simplifications presicely we are set on?

**Comment by vondele** [comment]
Let's go for my suggestion:  STC {-0.25, 1.25}, LTC {0.25, 1.25} and  STC {-1.25, 0.25} LTC {-0.75, 0.25}.

It would be great if you could make a PR!
This might be a starting point:
 https://github.com/glinscott/fishtest/pull/532
also wiki changes:
https://github.com/glinscott/fishtest/wiki/Creating-my-first-test#standard-tests-2
and adjusting the default draw rate of the SPRT calculator would be good as well...

**Comment by Vizvezdenec** [comment]
UGH...
I don't have fishtest fork on my PC, also I don't have any python compilers and stuff.
So for me it will be quite some work to do all of this and with my (almost) non-existent ability to use github it will take a lot of time... I will leave it to someone who actually can do it in reasonable timespan.

**Comment by vondele** [comment]
OK, so we can split the work :-) I can do the fishtest work, can you do the wiki ?

**Comment by Vizvezdenec** [comment]
Ah this is much easier part, np :)

**Comment by Vizvezdenec** [comment]
one question, what does RMS bias do and what value it should have? Currently default page leads to bias = 0 but fishtest wiki has it as 30. ?

**Comment by vondele** [comment]
IIUC, it is a property of the opening book (white advantage?), and only influences the number of games. Seemingly 30 is close enough to zero.

**Comment by Vizvezdenec** [comment]
Bounds: [-0.5,1.5]; draw ratio: 0.61; bias: 30.00; TC: 10+0.1; duration: 40 moves; time consumed: 0.90.
So I change it to 
Bounds: {-0.25, 1.25}; draw ratio: 0.78; bias: ?; TC: 10+0.1; duration: 40 moves; time consumed: ?.
So, first one should be 0? And what is time consumed? Still 0.9?

**Comment by vdbergh** [comment]
For noob_3moves the RMSbias is 30. The value is given at the bottom of the raw statistics page. 30 is quite small so there is not a big influence on the results.

**Comment by vondele** [comment]
TC: 10+0.1; duration: 40 moves; time consumed: 0.90. -> estimate of the time per game (10 + 40 * 0.1) * 2 * 0.9 
The more accurate estimate is now (10 + 68 * 0.1) * 2 * 0.92 see 
https://github.com/glinscott/fishtest/blob/aaf6b2926fb24b0b86f04dfb46d41ce77db43270/fishtest/fishtest/util.py#L233-L260

**Comment by Vizvezdenec** [comment]
shouldn't estimated number of moves be different for STC and LTC? I would assume that on average LTCs take more moves.
But in wiki they are the same. So I will just change moves number to 68 and time consumed to 0.92

**Comment by vondele** [comment]
yes, it is an estimate only. The difference wasn't large IIRC.

**Comment by Vizvezdenec** [comment]
Okay, I think I edited everything properly :)

**Comment by vondele** [comment]
photo finish...

**Comment by Vizvezdenec** [comment]
Okay, I'm closing this for now I guess. Let's see what will we get... :)

**Comment by MJZ1977** [comment]
I am already worried about queue length with new bounds. I thought that we will decrease TC at the same time. But lets see, I think we will need a lot of patience and discipline in next weeks :D.

**Comment by vondele** [comment]
TC decrease (nps rescaling) will come soon after a release. Let's see how long that takes. But patience and discipline sound good to me as well ;-)

**Comment by Vizvezdenec** [comment]
So with sf 12 release can we hope for baseline nps change? :)
